module4/persistence/src/main/java/com/epam/esm/persistence/impl/CertificateDaoImpl.java

```diff
@@ -0,0 +1,51 @@
+package com.epam.esm.persistence.impl;
+
+import com.epam.esm.criteria_info.CertificateCriteriaInfo;
+import com.epam.esm.criteria_info.PageInfo;
+import com.epam.esm.entity.Certificate;
+import com.epam.esm.persistence.CertificateDao;
+import com.epam.esm.persistence.dataspecification.CertificateSpecification;
+import com.epam.esm.persistence.repository.CertificateRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@RequiredArgsConstructor
+@Repository
+public class CertificateDaoImpl implements CertificateDao {
```

Comment:
Why do you use DAO layer between service and repository? I don't find specific logic in this class.

Reply:
to make service layer independent from `Specifications`

Reply:
Done